src/core/packet.lua

```diff
@@ -75,15 +84,49 @@ local function freelist_nfree(freelist)
    return freelist.nfree
 end
 
-local max_packets = 1e6
+local function freelist_lock(freelist)
+   sync.lock(freelist.lock)
+end
+
+local function freelist_unlock(freelist)
+   sync.unlock(freelist.lock)
+end
+
 local packet_allocation_step = 1000
 local packets_allocated = 0
-local packets_fl = ffi.new("struct freelist", max_packets, 0, max_packets)
+local packets_fl = ffi.new("struct freelist")
+      packets_fl.max = max_packets
+local group_fl
+if not shm.exists("group/packets.freelist") then
+   group_fl = shm.create("group/packets.freelist", "struct freelist")
+   group_fl.max = max_packets
+else
+   group_fl = shm.open("group/packets.freelist", "struct freelist")
+end
+
+function rebalance_freelists ()
+   if freelist_nfree(packets_fl) > packets_allocated then
+      freelist_lock(group_fl)
+      while freelist_nfree(packets_fl) > packets_allocated
```

Comment:
It sure would be nice to be able to transfer batches of packets to the group freelist, instead of transferring them one by one.  In that case you could potentially even avoid the spinlock around the group freelist and use atomic ops instead.  Anyway, something to consider for the future, it's clear that this works for now :)